### ISSUE TITLE
[Proposal] Add Official BF support (as independent language)

### ISSUE THREAD

**Issue by DonaldKellett** [Codewars/codewars-runner-cli#395]
*Note: BF is a common acronym for* [Brainf**k](https://esolangs.org/wiki/Brainfuck)

**I know this sounds wild**, but ... 

### Current Situation

Currently, it is entirely possible to write a Kata where the solver has to write a suitable program in BF to pass the tests specified as long as the Kata author knows how to properly implement a bug-free BF interpreter to interpret the user's solution.  However, such Kata are currently counted as a completion in another language entirely (e.g. JavaScript, see [this Kata](https://www.codewars.com/kata/582c2dffd82a1ff03000000a/discuss) and a few others) despite the fact that the user actually needs to know how to program in BF *instead of* that other language that the Kata is written in in order to complete it.  This poses a few problems; for example, if said BF kata is currently available in JavaScript and I translate it into PHP (and it gets approved), did I really translate that Kata into PHP?  Or did I just merely copy the Kata using PHP syntax instead of JavaScript syntax?  And for users who have completed said Kata in both JavaScript and PHP, do they **actually** know either of those two languages, or do they just know BF and instead got free rank progress in both JavaScript and PHP?

### Rationale

The official support of BF as an independent language solves the logical dilemma above and gives the correct credit to the solver of the Kata as having completed a Kata in BF and not in those other languages.

### How To Support BF

Initially, perhaps we can add basic BF support in Kumite Only without a testing framework which I guess should be easy enough as it shouldn't require much more than implementing a working interpreter for it.  Then, in the (hopefully near) future, we could perhaps create a custom testing framework **which is written in another (popular) language such as JavaScript or Ruby** (like we did for SQL) designed to test BF code specifically.  I know that BF is technically "Turing complete" so it is *technically* possible to write tests and preloaded code entirely in BF **but** we don't want to scare people off by making tests insanely hard to write :P

### Additional Notes

Currently, Kata involving writing BF programs do not receive much attention at all from the Community (not a single one of the BF kata authored by [gkucmierz](http://codewars.com/users/gkucmierz/authored) have even come close to approval yet) but I think the problem is **not** because BF is unpopular, but rather because it is not officially credited by the Codewars system as the solver making rank progress in BF which in turn provides no motivation for BF enthusiasts to actually complete such Kata.

**Comment by Voileexperiments** [comment]
I'd suggest supporting arbitrary (esolang/domain-specific/custom) and even fictional languages as well: I'd love to see how people code in VHDL, or language in programming games like TIS-100/SHENZHEN I/O.

**Comment by DonaldKellett** [comment]
@Voileexperiments That would be an interesting feature to add but how would it count towards a user's rank progress?

**Comment by jhoffner** [comment]
I'm all for it but this is most definitely going to be needed to be contributed by someone in the community.

**Comment by DonaldKellett** [comment]
Thanks for your support @jhoffner , would you mind pointing me in the right direction regarding this?  For example, what files do I have to write, which folders they should belong to, whether command line is involved at any point in the process, etc.  I'll try to add BF support (at least for Kumite for now) as long as it's within my capabilities :smile:

**Comment by jhoffner** [comment]
Sure @DonaldKellett. I would recommend reading the entire readme but here are some instructions on adding a new language. https://github.com/Codewars/codewars-runner-cli#how-to-add-a-new-language

The command line is involved, as well as Docker. You would need to install docker and then I would recommend using the docker-compose based examples as opposed to using it directly. 

You could reference the php and javascript runner implementations to help get your bearings. Also check out @kazk's recent PR's for new languages as they tend to be very complete examples of adding language support. Here is an [example](https://github.com/Codewars/codewars-runner-cli/pull/310/files)

**Comment by kazk** [comment]
@DonaldKellett this should be straight forward and shouldn't take long even if you're not familiar with the codebase :)

Something like this:

- Find an implementation to use ([bf](http://packages.ubuntu.com/source/trusty/bf), [beef](http://packages.ubuntu.com/source/trusty/beef), etc.)
- Write `docker/esolangs.docker` and install them
- Write `lib/runners/bf.js` to do `bf/beef [options] code.b`
- Write `test/runners/bf_spec.js` to check if it works
- Modify `Makefile`
  - Add to `esolangs` to `CONTAINERS` list
  - Set `RECENT_CONTAINERS`

Before you start, I'd recommend you to learn `git` because you're still doing

![image](https://cloud.githubusercontent.com/assets/639336/25366999/0c895ad0-2928-11e7-9c66-578aa2075cd2.png)

and work on `feature/bf` instead of `master`.

---

@jhoffner how about `language: esolang` and `languageVersion: bf`? This will enable what @Voileexperiments commented above. Tests can be done similar to `shell`.

**Comment by DonaldKellett** [comment]
Thanks a lot for the instructions @jhoffner and @kazk :smile:  I currently don't know Docker and struggle often with Git (e.g. receiving errors when I attempt to sync my codewars-runner-cli fork with this official Repo using Git directly and not knowing why) but I'll try to learn some basic Docker over the next few days/weeks by viewing online tutorials and see if I can complete the task at hand.

@gkucmierz Seeing as you appear to still be active on Codewars and that you authored a bunch of BF Kata a few months back, would you by any chance be interested to take on this Task on making BF an official language on Codewars?

**Comment by kazk** [comment]
@DonaldKellett I think this will be a nice task to familiarize yourself with this codebase so it's worth a try. I'll try to help if you get stuck. If you don't want to do it, I can do it.

**Comment by DonaldKellett** [comment]
@kazk I'll try to learn some basic Docker within these few days.  And then, *if* I fail to create anything of use using Docker even after learning the basics, I'll ping (i.e. mention) you again and perhaps you can take over if it doesn't distract you from tasks that are higher up the priority list on this Runner?  Thanks a lot again :smile:

But before I even start with Docker, I think I'll have to retake the crash course on Git Basics on Codecademy just to familiarize myself with terms such as `origin` and `target` just to fix my Issue of making loads of unnecessary merge commits 😛  Off I go ...

**Comment by DonaldKellett** [comment]
@kazk I just re-took the Git Crash Course on Codecademy and retried the command line example you gave me last time regarding removing all those unnecessary merge commits which worked this time so I would like to thank you again for your help :smile:  I'm now planning to watch some introductory videos on Docker before I delve into the codebase and hopefully be able to contribute something useful towards adding BF as an officially supported language.

![2017-04-26 1 52 18](https://cloud.githubusercontent.com/assets/7777884/25419840/9fb12594-2a87-11e7-949e-47e2888922e3.png)

**Comment by DonaldKellett** [comment]
@kazk I've recently managed to follow through all but the final section (actually deploying the app) of the official [Get Started](https://docs.docker.com/get-started/) provided by Docker.  I've just re-read your directions on how to set up Docker for BF and I have a question that I would like to ask:

What does `docker/esolangs.docker` do?  I've followed the basic tutorial and have heard of `Dockerfile`, `requirements.txt`, `docker-compose.yml` and the like but I haven't worked with any files that ended with `.docker` yet.  Is that file (`esolangs.docker`) similar to `requirements.txt` in terms of purpose or am I gravely mistaken?  If it is indeed similar to `requirements.txt` in that official Tutorial, do I just list `bf` (or `beef`) as a package to be installed there?

**Comment by kazk** [comment]
> What does `docker/esolangs.docker` do?

It's just a `Dockerfile`. `docker/esolangs.docker` will be copied to the root of the project as `Dockerfile` when building the image. This is currently necessary because of the way things are setup, having multiple `Dockerfiles` in one place.

`requirements.txt` in that tutorial is used by `pip install -r` to install *python packages* so don't worry about that.

**Comment by kazk** [comment]
@DonaldKellett there's something weird going on with your fork.

Your fork is adding another set of tests in [javascript_spec.js#L790-L870](https://github.com/DonaldKellett/codewars-runner-cli/blob/3421a7a58641982f3ef9f60d8e8a6590babffd75/test/runners/javascript_spec.js#L790-L870). The tests you added are already in [javascript_spec.js#L708-L788](https://github.com/DonaldKellett/codewars-runner-cli/blob/3421a7a58641982f3ef9f60d8e8a6590babffd75/test/runners/javascript_spec.js#L708-L788).

![image](https://cloud.githubusercontent.com/assets/639336/25464121/8631d078-2aaf-11e7-9592-fe6e5925ffa1.png)

![image](https://cloud.githubusercontent.com/assets/639336/25464141/9b00d846-2aaf-11e7-8d1a-1ceadcdb2495.png)

The commands I suggested to you in #369 was for getting rid of unnecessary merge commits and *applying* the changes you had made to the latest upstream `master` at the time. Executing those commands exactly won't work anymore because the changes were already applied by merging the PR.

**Comment by DonaldKellett** [comment]
Thanks @kazk for the heads up regarding the duplicate lines of code; I am now proceeding to remove lines 790 to 872 in `javascript_spec.js` in both my `feature/bf` and `master` branches.  Regarding the `docker/esolangs.docker`, many thanks for your explanation, I'll try to crack on with it later today :smile:

**Comment by kazk** [comment]
> I am now proceeding to remove lines 790 to 872 in javascript_spec.js in both my feature/bf and master branches.

Make sure to reset/revert instead of deleting the lines manually and committing.

> Regarding the docker/esolangs.docker, many thanks for your explanation

No problem. I made a proposal to modularize this codebase few days ago (#396) and one of the reasons is to lower the barrier of contribution. So it's good to know what's confusing for contributors like you.

**Comment by kazk** [comment]
@DonaldKellett maybe it's more simple to just delete the fork and fork it again. I only noticed the last commit, but there might be more weirdness in the history of your fork.

**Comment by DonaldKellett** [comment]
Thanks for the suggestion, I've just re-forked the entire Repo.  Why didn't I think of it before? 😛

Will be re-creating the `feature/bf` branch and experimenting now.

**Comment by DonaldKellett** [comment]
@kazk I've just studied a few of those Dockerfiles in the `docker/` directory in an attempt to write my own `esolangs.docker` and so far I've noticed a few things:

1. All Dockerfiles start with pulling the base image
2. There must be installation of some language-related packages (or similar)
3. The cli-runner is always added
4. There are at least 2 environmental variables `HOME` and `USER` which always take the same values
5. The final line is always the one handling potential node errors

Before I actually write my file, may you please check this draft and tell me how close it is to what is required?

```docker
# Pull base image.
FROM codewars/base-runner

# Install "bf" Fast Brainf**k Interpreter (I'm not very sure about the syntax of this one)
RUN install bf # ???

# ADD cli-runner and install node deps
ADD . /runner

RUN ln -s /home/codewarrior /workspace

# Set environment variables
ENV HOME /home/codewarrior
USER codewarrior
ENV USER codewarrior

#timeout is a fallback in case an error with node
#prevents it from exiting properly
ENTRYPOINT ["timeout", "15", "node"]
```

**Comment by kazk** [comment]
You got the basic outline correct :+1:

```dockerfile
# Install "bf" Fast Brainf**k Interpreter
RUN apt-get update && apt-get install -y --no-install-recommends bf
```

For the detailed explanation, see [explainshell.com](https://explainshell.com/explain?cmd=apt-get+update+%26%26+apt-get+install+-y+--no-install-recommends+bf).
I think it will help you a lot getting used to the command line :)

![image](https://cloud.githubusercontent.com/assets/639336/25469462/213582e2-2ad1-11e7-83d2-54154b90233d.png)

**Comment by DonaldKellett** [comment]
> - Write `docker/esolangs.docker` and install them

I've just added and written a new file `esolangs.docker` in the `docker` directory of my `feature/bf` branch of the repository and I now have my Terminal ready with Docker on.  Once I change directories to my local copy of `codewars-runner-cli`, what Docker commands should I execute to install them?  I was initially thinking changing directory again to the `docker/` directory and do something like `docker build -t esolangs .` but then I realised that there is more than one Dockerfile in that directory and I didn't know how to choose `esolangs.docker` specifically to build my image.  In case it makes any difference, I've navigated to my `codewars-runner-cli/` directory and successfully executed `make node` right after I turned on Docker and before I typed this comment.

BTW @kazk thanks for the site recommendation (on explaining Shell), it was a very good read :smile: :+1:

**Comment by kazk** [comment]
> what Docker commands should I execute to install them?

Have you added `esolangs` to `Makefile`? By adding it there like other images, you'll be able to do `make esolangs` to build it just like you did `make node`.

`make esolangs` will do the following commands for you:

```
cp docker/esolangs.docker ./Dockerfile
docker build -t codewars/esolangs-runner .
```

**Comment by DonaldKellett** [comment]
> Have you added `esolangs` to `Makefile`? ... 

Oh, I thought this was the last step but I'll do it first anyway :smile:

**Comment by DonaldKellett** [comment]
So now I think I've done the first, second and final bullet points of your original TODO list, and I'm on the third item which is:

> - Write `lib/runners/bf.js` to do `bf/beef [options] code.b`

I've just skimmed through a few examples in the `lib/runners/` directory and I've noticed that all of them require `shovel` (`require('../shovel')`) and most of them also require `path`.  In terms of the shovel, I'm currently only planning to add support for basic run so I assume the object passed in as the 3rd argument of `shovel.start` should only have the `solutionOnly` method?  This is currently what I am thinking for `lib/runners/bf.js` but would you mind reviewing it for me again @kazk before I add this file to my fork and make a commit?

```javascript
var shovel = require('../shovel');

module.exports.run = function run(opts, cb) {
  shovel.start(opts, cb, {
    solutionOnly: function(runCode) {
      runCode({
        name: 'bf',
        args: [[opts.setup ? opts.setup : '', opts.solution].join('\n'), 'code.b']
      });
    }
  });
};
```

Many thanks for your help and guidance again :smile:

**Comment by kazk** [comment]
You need to write the bf code to file.
You can use `writeFileSync` from `lib/util.js` to do that, which returns the file path. Then you can pass that path to `bf` command.
Also, other runners without test integration still provides `testIntegration` which throws an error.

> This is currently what I am thinking for `lib/runners/bf.js` but would you mind reviewing it for me again @kazk before I add this file to my fork and make a commit?

? I don't understand what you mean by "add this file to my fork and make a commit".
You should write a test and see why it fails. If you can't figure it out, I'll try to help :)

**Comment by DonaldKellett** [comment]
> You should write a test and see why it fails. If you can't figure it out, I'll try to help :)

Right.  Sorry, I was being lazy 😛

I just read a few other runners in detail and so far, it appears that `util.codeWriteSync` is used far more often than `util.writeFileSync`.  So what's the difference between the two methods (apart from their argument list of course) in terms of what they accomplish?  Or do they both write code into a specified file in a specified path?

**Comment by DonaldKellett** [comment]
@kazk So now I think I've taken a stab at all of the bullet points you listed for adding Esolangs (BF specifically) support and I executed `make esolangs` again on my Docker.  All of the commands executed just fine except for this one (which I adapted from the `node.docker` to run the BF spec tests):

```docker
RUN mocha -t 10000 test/runners/bf_spec.js
```

Instead of passing/failing tests, the command line gave me this error (despite me having added a file `bf_spec.js` in the `test/runners/` directory):

```
Warning: Could not find any test files matching pattern: test/runners/bf_spec.js
No test files found
```

AFAIK I haven't misspelled anything.  Any idea why that might occur and how to fix it?

By the way, I'm sure I made a commit after adding both the `bf_spec.js` and the line of code shown above and synchronized it with my online GitHub copy so that can't be the root of the problem.  Thanks a lot :smile:

**Comment by kazk** [comment]
@DonaldKellett I think you're missing

```dockerfile
WORKDIR /runner
```

Try

```dockerfile
WORKDIR /runner
RUN mocha -t 10000 test/runners/bf_spec.js
```

---

Also, you should consider squashing your commits before opening a PR.

**Comment by DonaldKellett** [comment]
Thanks @kazk , that appears to have solved the error I was getting previously, but now I'm getting an error when attempting `require('chai')` that the module cannot be found.  So far, I have searched a few other Spec files and they all require "chai" in the exact same way.  I've also checked any code below those files and it appears that all of the variable declarations are in the top and the rest are `describe` and `it` blocks.  Do you know what might be causing that problem?

> Also, you should consider squashing your commits before opening a PR.

I'm not sure how I can do that?  If I first reset my HEAD to a previous commit then I might lose my changes in the working directory.

**Comment by kazk** [comment]
Try adding

```dockerfile
# add the package json first to a tmp directory and build, copy over so that we don't rebuild every time
ADD package.json /tmp/package.json
RUN cd /tmp && npm install --production
RUN mkdir -p /runner && cp -a /tmp/node_modules /runner
```

before

```dockerfile
ADD . /runner
```

---

For squashing, I usually do `git rebase -i HEAD~n`, but you should go read about them because I don't have the time to explain in detail and it's out of scope.
I wrote "consider" because some projects don't care. Maybe ask @jhoffner if you think it's unnecessary.
I *personally* think commits should be grouped into meaningful changes so the history is useful. I also do it to make them easy to review.
My **local** commits *sometimes* looks like:

```
[WIP] Adding X
[WIP] fix Y
[WIP] revert Y
[WIP] ...wtf
[WIP] should work now
```

But I make it:

```
Add X
```

when it's done before pushing :)

**Comment by DonaldKellett** [comment]
@kazk Many thanks yet once again, I've added the lines of code you suggested and it indeed solved the Issue.  I was especially glad to see that the actual tests passed as well on the first run.

With regards to squashing the commits on my branch, I searched up the command on ExplainShell and read an article on `git rebase` from the official Git website but it currently does not appear to be very successful.  I couldn't figure out a way of continuing the process after I have entered Interactive Mode and made the necessary edits, and when I tried the other variants (e.g. `git rebase --onto somebranch~n somebranch~m somebranch`), even when they worked locally, the changes all disappeared once I clicked "Sync".  On top of that, I made a wrong move and almost undid around 10 commits.  So for now I think I am going to stay on the safe side and hope Jake doesn't mind all those extra commits I added 😛

The pull request is now open at https://github.com/Codewars/codewars-runner-cli/pull/405 so I think I can close this Issue now.  If my pull request isn't successful and is rejected as a result, perhaps I may re-open this and attempt another pull request.
